salt/transport/tcp.py

```diff
@@ -773,38 +812,200 @@ def handle_future(future):
+                )
+
+    @tornado.gen.coroutine
+    def _stream_read(self, client):
+        unpacker = msgpack.Unpacker()
+        while not self._closing:
+            try:
+                client._read_until_future = client.stream.read_bytes(4096, partial=True)
+                wire_bytes = yield client._read_until_future
+                unpacker.feed(wire_bytes)
+                for framed_msg in unpacker:
+                    body = framed_msg['body']
+                    if body['enc'] != 'aes':
+                        # We only accept 'aes' encoded messages for 'id'
+                        continue
+                    crypticle = salt.crypt.Crypticle(self.opts, salt.master.SMaster.secrets['aes']['secret'].value)
+                    load = crypticle.loads(body['load'])
+                    if not self.aes_funcs.verify_minion(load['id'], load['tok']):
+                        continue
+                    client.id_ = load['id']
+                    self._add_client_present(client)
+            except tornado.iostream.StreamClosedError as e:
+                log.debug('tcp stream to {0} closed, unable to recv'.format(client.address))
+                client.close()
+                self._remove_client_present(client)
+                self.clients.discard(client)
+                return
+            except Exception as e:
+                log.error('Exception parsing response', exc_info=True)
+                continue
 
     def handle_stream(self, stream, address):
         log.trace('Subscriber at {0} connected'.format(address))
-        self.clients.append((stream, address))
+        client = PubServerClient(stream, address)
+        self.clients.add(client)
+        self.io_loop.spawn_callback(self._stream_read, client)
 
     # TODO: ACK the publish through IPC
     @tornado.gen.coroutine
-    def publish_payload(self, payload, _):
-        log.debug('TCP PubServer sending payload: {0}'.format(payload))
-        payload = salt.transport.frame.frame_msg(payload['payload'])
+    def publish_payload(self, package, _):
+        log.debug('TCP PubServer sending payload: {0}'.format(package))
+        payload = salt.transport.frame.frame_msg(package['payload'])
 
         to_remove = []
-        for item in self.clients:
-            client, address = item
-            try:
-                # Write the packed str
-                f = client.write(payload)
-                self.io_loop.add_future(f, lambda f: True)
-            except tornado.iostream.StreamClosedError:
-                to_remove.append(item)
-        for item in to_remove:
-            client, address = item
-            log.debug('Subscriber at {0} has disconnected from publisher'.format(address))
+        if 'topic_lst' in package:
+            topic_lst = package['topic_lst']
+            for topic in topic_lst:
+                if topic in self.present:
+                    # Create a separate copy (as a list this time) to iterate
```

Comment:
Since the publisher runs in a single process, and this next block of code has no `yield` statements, you shouldn't have to make the copy since you are the only coroutine running until you `yield`.